concepts/infrastructure-reconciliation/README.md

```diff
@@ -1,41 +1,196 @@
+
+### Issuing RBAC-based kubeconfigs
+
+To provide an environment, where certain Operators are not able to leverage the cluster access, it is required to open up a possibility for the constrained kubeconfigs to be created. This ensures that Operators can work only with the resources they require access to and security-wise, potential malicious parties won't be able to use kubeconfigs issued for the Operators to execute cluster-wide attacks.
+
+The solution is to introduce a possibility in the Cluster Managers to issue limited kubeconfigs, which will be rotated in a time-based manner. Creation of those resources in Kyma Runtimes will be done using the master kubeconfigs coming from the Gardener (or Vault for BYOC), which will not be stored anywhere in the Control Plane.
+
+### Dependency management
+
+It should be the responsibility of the Component Reconcilers to figure out if all dependencies exist and the reconciliation can be completed. This is true for all the infrastructure-related artefacts, such as the cluster state and kubeconfigs. If dependencies are not ready, the Component Reconciler should return the information that it is not ready and wait for the next trigger (time or event).
+
+## High-level architecture
 
 ![](assets/infrastructure-reconciliation.svg)
 
-Mothership Reconciler should provide common functionality related to watching changes and triggering Component Reconcilers in response to events or with the time-based schedule. Such aspects as queueing reconciliation requests, consolidating them, updating reconciliation status, retries and back-off strategy should be implemented once (in the Mothership). 
+The three new operators would need to be introduced:
+- Gardener Cluster Manager
+- Customer Cluster Manager
+- Subscription Manager
+
+Each of those components has a clear set of responsibilities that together handle the flow that creates a new, complete, Kyma-ready  Kubernetes cluster.
+
+Along the operators, there are four new Custom Resources, that serve as an API for the components:
+- Cluster CR
+- Gardener Cluster CR
+- Customer Cluster CR
+- SubscriptionRequest CR
+
+### Cluster Managers and the Cluster CRs
+
+The Cluster Managers' main responsibility is to manage the respective Cluster CRs, this is tied to both their provisioning and providing an access to the created K8S clusters.
+
+First of all, an overarching Cluster CR will provide a common API for operations connected to support or infrastructure maintenance. It contains reference to the resource corresponding to the sub-type of cluster:
+```
+apiVersion: operator.kyma-project.io/v1alpha1
+kind: Cluster
```

Comment:
I think this kind of cluster API is bound to not cover all cases and we should not introduce a common Cluster CR if we anyhow have two separate CRs (GardenerCluster and CustomerCluster). To be frank, all this does is hold a ref to another CR, so why even reconcile it instead of reconciling Customer/Gardener Clusters directly?

Reply:
Also, if we want to keep Cluster as general API, I think this is an incorrect implementation since you bind the Interface API (Kind Cluster) with different Ref types to their implementations (gardenerClusterRef and customerClusterRef). Looking at Kubernetes upstream, this mistake was done multiple times in API Design (e.g. in the design phase of PVs and StorageClass) and they had to redo this.

Looking at the learnings, I think we should follow a similar approach to the upcoming Kubernetes Gateway API (see https://gateway-api.sigs.k8s.io/api-types/gatewayclass/):

This would mean that we have 1 ClusterClass for each of the implementations and an Interface API that can deal with interpreting calls from a ClusterClass. Then the Cluster Controller Implementations only have to be able to deal with the ClusterClass implementation by using a field called "clusterClassName: Gardener" in the Spec of the Cluster. That way you do not have to introduce new Ref Types for a new CRD you want to add as implementation detail.

Let me know what you think!

Reply:
The goal was to not reconcile the common Cluster CR at all - rather keep it as a way to simplify the operations regarding environment management, support or statistics. The operators will reconcile only their respective cluster types respectively (ExternalCluster or GardenerCluster).

I believe it can be dropped at all - for now, there is no use case where the common API would be required (AFAIK). It can be easily worked around, the scripts and other potential automation would just need to account for both types of clusters. Perhaps @pbochynski or @ebensom may have more insights on the requirements here.

Reply:
I agree - we can postpone creating generic Cluster CR when we have clear demand for it.